docs/security-considerations.rst

```diff
@@ -344,3 +346,176 @@ Note that formal verification itself can only help you understand the
 (the actual implementation). You still need to check whether the specification
 is what you wanted and that you did not miss any unintended effects of it.
+
+Solidity implements a formal verification approach based on SMT solving.  The
+SMTChecker module automatically tries to prove that the code satisfies the
+specification given by ``require/assert`` statements, that is, it considers
+``require`` statements as assumptions and tries to prove that the conditions
+inside ``assert`` statements are always true.  If an assertion failure is
+found, a counterexample is given to the user, showing how the assertion can be
+violated.
+
+The SMTChecker also checks automatically for underflow/overflow, trivial
+conditions and unreachable code.
+It is currently an experimental feature, therefore in order to use it you need
+to enable it via :ref:`a pragma directive<smt_checker>`.
+
+The SMTChecker traverses the Solidity AST creating and collecting program constraints.
+When it encounters a verification target, an SMT solver is invoked to determine the outcome.
+
+For more details on how the SMT encoding works internally, see the paper
+`SMT-based Verification of Solidity Smart Contracts <https://github.com/leonardoalt/text/blob/master/solidity_isola_2018/main.pdf>`_.
+
+Abstraction and false positives
+===============================
+
+The SMTChecker implements abstractions in an incomplete and sound way: If a bug
+is reported, it might be a false positive introduced by abstractions (while
+erasing knowledge or using a non-precise type). If it determines that a
+verification target is safe, it is indeed safe, that is, there are no false
+negatives (unless there is a bug in the SMTChecker).
+
+The SMT encoding tries to be as precise as possible, mapping Solidity types
+and expressions to their closest `SMT-LIB <http://smtlib.cs.uiowa.edu/>`_
+representation, as shown in the table below.
+
++-----------------------+--------------+-----------------------------+
+|Solidity type          |SMT sort      |Theories (quantifier-free)   |
++=======================+==============+=============================+
+|Boolean                |Bool          |Bool                         |
++-----------------------+--------------+-----------------------------+
+|intN, uintN, address,  |Integer       |LIA, NIA                     |
+|bytesN, enum           |              |                             |
++-----------------------+--------------+-----------------------------+
+|array, mapping         |Array         |Arrays                       |
++-----------------------+--------------+-----------------------------+
+|other types            |Integer       |LIA                          |
++-----------------------+--------------+-----------------------------+
+
+Types that are not yet supported are abstracted by a single 256-bit unsigned integer,
+where their unsupported operations are ignored.
+
+Function calls are inlined when possible, that is, when their implementation is
```

Comment:
I guess even if the implementation of external functions is available they are not inlined (because of polymorphism), is that the case?

Reply:
Ah, it's not the case. Hm. I guess that might be fine, especially because we can also have polymorphism with internal functions. Might be dangerous, though, if someone does not use proper interfaces with no implementation.

Reply:
At the moment the SMTChecker only inlines calls to internal functions.

Even though we can also have polymorphism with internal functions, they are resolved at compile time, right? So it'd be safe to inline.
We can also decide against inlining external functions where the code is not resolved in compile time.

Reply:
Yes, internal polymorphism is resolved at compile time. Do you visit all functions in the current contract or also all functions in all base contracts?

Reply:
Only in the current contract.